File: sys/petscOptionsSetInt.m
function [errCode, toplevel] = petscOptionsSetInt(iname, value)
% Sets an option name-value pair into the options database, overriding
% whatever is already present.
%
%   errCode = petscOptionsSetInt(iname, value) sets the option
%   iname in the option database to a given integer value. The iname must start
%   with '-', and must be null-terminated.
%
%   errCode(int) return code (0 indicates OK)
%
% This is a high-level wrapper for petscOptionsSetValue.
%
% SEE ALSO: petscOptionsSetReal, petscOptionsSetValue


%#codegen -args {coder.typeof(char(0), [1,inf]), int32(0)}

errCode = int32(-1);

if ~isempty(coder.target)
    toplevel = nargout>1;
    if ~isempty(iname) && iname(end) && (toplevel || m2c_debug)
        m2c_error('MPETSc:petscOptionsSetInt:InputError', ...
            'Argument name must be a null-terminated string.')
    end
    
    str = m2c_int2str(int32(value));
    errCode = coder.ceval('PetscOptionsSetValue', PETSC_NULL_OPTIONS, ...
        coder.rref(iname), coder.rref(str));
    
    if errCode && (toplevel || m2c_debug)
        m2c_error('petsc:RuntimeError', 'PetscOptionsSetValue returned error code %d\n', errCode)
    end
end
end

File: sys/codegen/lib/petscOptionsSetInt/petscOptionsSetInt.c
#include "petscOptionsSetInt.h"
#include "petscOptionsSetInt_types.h"
#include "m2c.h"
#include "petsc4m.h"

static void b_m2c_error(int varargin_3);

static void m2c_error(void);

static void b_m2c_error(int varargin_3)
{
  const char *fmt;
  const char *msgid;
  msgid = "petsc:RuntimeError";
  fmt = "PetscOptionsSetValue returned error code %d\n";
  M2C_error(msgid, fmt, varargin_3);
}

static void m2c_error(void)
{
  const char *fmt;
  const char *msgid;
  msgid = "Petsc4m:petscOptionsSetInt:InputError";
  fmt = "Argument name must be a null-terminated string.";
  M2C_error(msgid, fmt);
}

void petscOptionsSetInt(const emxArray_char_T *iname, int value, int *errCode,
                        boolean_T *toplevel)
{
  PetscOptions obj;
  int i;
  char str[20];
  signed char t_str[20];
  char *ptr;
  *toplevel = true;
  if ((iname->size[1] != 0) && (iname->data[iname->size[1] - 1] != '\x00')) {
    m2c_error();
  }
  ptr = (char *)(&t_str[0]);
  sprintf(ptr, "%d", value);
  for (i = 0; i < 20; i++) {
    str[i] = t_str[i];
  }
  obj = NULL;
  *errCode = PetscOptionsSetValue(obj, &iname->data[0], &str[0]);
  if (*errCode != 0) {
    b_m2c_error(*errCode);
  }
}

void petscOptionsSetInt_initialize(void)
{
}

void petscOptionsSetInt_terminate(void)
{
}
